packages/engine/Source/Core/Event.js

```diff
@@ -128,29 +119,24 @@ function compareNumber(a, b) {
 Event.prototype.raiseEvent = function () {
   this._insideRaiseEvent = true;
 
-  let i;
-  const listeners = this._listeners;
-  const scopes = this._scopes;
-  let length = listeners.length;
-
-  for (i = 0; i < length; i++) {
-    const listener = listeners[i];
-    if (defined(listener)) {
-      listeners[i].apply(scopes[i], arguments);
+  for (const [listener, scopes] of this._listeners.entries()) {
+    for (const scope of scopes) {
+      if (defined(listener)) {
+        listener.apply(scope, arguments);
+      }
```

Comment:
If the listener is undefined, we can skip iterating over the scopes at all, right?

```suggestion
  for (const [listener, scopes] of this._listeners.entries()) {
    if (!defined(listener)) {
      continue;
    }
    for (const scope of scopes) {
      listener.apply(scope, arguments);
    }
```

Reply:
Good point!